src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.cs

```diff
@@ -0,0 +1,228 @@
+            {
+                context.RegisterOperationBlockStartAction(OnOperationBlock);
+                context.RegisterSymbolEndAction(OnSymbolEnd);
+            }
+
+            private void OnOperationBlock(OperationBlockStartAnalysisContext context)
+            {
+                context.RegisterOperationAction(OnMethodReference, OperationKind.MethodReference);
+                BlockAnalyzer.Analyze(context, this);
+            }
+
+            private void OnMethodReference(OperationAnalysisContext context)
+            {
+                var methodBinding = (IMethodReferenceOperation)context.Operation;
+                _methodsUsedAsDelegates.GetOrAdd(methodBinding.Method.OriginalDefinition, true);
+            }
+
+            private void OnSymbolEnd(SymbolAnalysisContext context)
+            {
+                foreach (var parameterAndUsageKvp in _unusedParameters)
+                {
+                    var parameter = parameterAndUsageKvp.Key;
+                    bool hasReference = parameterAndUsageKvp.Value;
+
+                    ReportUnusedParameterDiagnostic(parameter, hasReference, context.ReportDiagnostic, context.Options, context.CancellationToken);
+                }
+            }
+
+            private void ReportUnusedParameterDiagnostic(
+                IParameterSymbol parameter,
+                bool hasReference,
+                Action<Diagnostic> reportDiagnostic,
+                AnalyzerOptions analyzerOptions,
+                CancellationToken cancellationToken)
+            {
+                if (!IsUnusedParameterCandidate(parameter))
+                {
+                    return;
+                }
+
+                var location = parameter.Locations[0];
+                var optionSet = analyzerOptions.GetDocumentOptionSetAsync(location.SourceTree, cancellationToken).GetAwaiter().GetResult();
+                if (optionSet == null)
+                {
+                    return;
+                }
+
+                var option = optionSet.GetOption(CodeStyleOptions.UnusedParameters, parameter.Language);
+                if (option.Notification.Severity == ReportDiagnostic.Suppress ||
+                    !ShouldReportUnusedParameters(parameter.ContainingSymbol, option.Value, option.Notification.Severity))
+                {
+                    return;
+                }
+
+                var diagnostic = DiagnosticHelper.CreateWithMessage(s_unusedParameterRule, location,
+                    option.Notification.Severity, additionalLocations: null, properties: null, message: GetMessage());
+                reportDiagnostic(diagnostic);
+
+                return;
+
+                // Local functions.
+                LocalizableString GetMessage()
+                {
+                    LocalizableString messageFormat;
+                    if (parameter.ContainingSymbol.HasPublicResultantVisibility() &&
+                        !parameter.ContainingSymbol.IsLocalFunction())
```

Comment:
capturing from a distance.    i would prefer this passed into GetMessage to make it clear at callsite this is dependent on these values.

Reply:
Refactored this method.

---
In reply to: [234868516](https://github.com/dotnet/roslyn/pull/30777#discussion_r234868516) [](ancestors = 234868516)